File: fresher_food_backend/FressFood/Models/Coupon.cs
namespace FressFood.Models
{
    public class Coupon
    {
        public string Id_phieugiamgia { get; set; } = string.Empty;
        public string Code { get; set; } = string.Empty;
        public decimal GiaTri { get; set; }
        public string? MoTa { get; set; }
        public string LoaiGiaTri { get; set; } = "Amount"; // "Amount" hoặc "Percent"
        public int? SoLuongToiDa { get; set; } // Số lượng tối đa có thể sử dụng (NULL = không giới hạn)
        public int SoLuongDaSuDung { get; set; } = 0; // Số lượng đã sử dụng
    }
}

File: fresher_food/lib/models/Coupon.dart
class PhieuGiamGia {
  final String idPhieuGiamGia;
  final String code;
  final double giaTri;
  final String moTa;
  final String loaiGiaTri; // "Amount" hoặc "Percent"
  final int? soLuongToiDa; // Số lượng tối đa có thể sử dụng (null = không giới hạn)
  final int soLuongDaSuDung; // Số lượng đã sử dụng

  const PhieuGiamGia({
    required this.idPhieuGiamGia,
    required this.code,
    required this.giaTri,
    required this.moTa,
    this.loaiGiaTri = 'Amount', // Mặc định là số tiền cố định
    this.soLuongToiDa,
    this.soLuongDaSuDung = 0,
  });

  factory PhieuGiamGia.fromJson(Map<String, dynamic> json) {
    return PhieuGiamGia(
      idPhieuGiamGia: json['id_phieugiamgia']?.toString() ?? '',
      code: json['code']?.toString() ?? '',
      giaTri: (json['giaTri'] is num) ? (json['giaTri'] as num).toDouble() : 0.0,
      moTa: json['moTa']?.toString() ?? '',
      loaiGiaTri: json['loaiGiaTri']?.toString() ?? 'Amount',
      soLuongToiDa: json['soLuongToiDa'] != null ? int.tryParse(json['soLuongToiDa'].toString()) : null,
      soLuongDaSuDung: json['soLuongDaSuDung'] != null ? int.tryParse(json['soLuongDaSuDung'].toString()) ?? 0 : 0,
    );
  }

  // Kiểm tra xem voucher còn có thể sử dụng không
  bool get conSuDungDuoc {
    if (soLuongToiDa == null) return true; // Không giới hạn
    return soLuongDaSuDung < soLuongToiDa!;
  }
}